iso8601/parse.go

```diff
@@ -0,0 +1,128 @@
+package iso8601
+
+import (
+	"encoding/binary"
+	"time"
+	"unsafe"
+)
+
+// Parse parses an ISO8601 string, e.g. "2021-03-25T21:36:12Z".
+func Parse(input string) (time.Time, error) {
+	switch b := unsafeStringToBytes(input); len(b) {
+	case 20:
+		t1 := binary.LittleEndian.Uint64(b)
+		t2 := binary.LittleEndian.Uint64(b[8:])
+		t3 := uint64(binary.LittleEndian.Uint32(b[16:]))
+
+		// Check for valid "YYYY-MM-DDTHH:MM:SSZ" separators by masking input
+		// with "    -  -  T  :  :  Z". If separators are all valid, replace
+		// them with a '0' (0x30) byte and check all bytes are now numeric.
+		// If it's not input we can handle on the fast path, pass it off to
+		// time.Parse().
+		if !match(t1, mask1) || !match(t2, mask2) || !match(t3, mask3) {
+			goto fallback
+		}
+		t1 ^= replace1
+		t2 ^= replace2
+		t3 ^= replace3
+		if (nonNumeric(t1) | nonNumeric(t2) | nonNumeric(t3)) != 0 {
+			goto fallback
+		}
+
+		t1 -= zero
+		t2 -= zero
+		t3 -= zero
+		year := (t1&0xF)*1000 + (t1>>8&0xF)*100 + (t1>>16&0xF)*10 + (t1 >> 24 & 0xF)
+		month := (t1>>40&0xF)*10 + (t1 >> 48 & 0xF)
+		day := (t2&0xF)*10 + (t2 >> 8 & 0xF)
+		hour := (t2>>24&0xF)*10 + (t2 >> 32 & 0xF)
+		minute := (t2>>48&0xF)*10 + (t2 >> 56)
+		second := (t3>>8&0xF)*10 + (t3 >> 16)
+
+		if month > 12 || day > 31 || hour >= 24 || minute >= 60 || second >= 60 {
+			goto fallback
+		} else if month == 2 && (day > 29 || (day == 29 && !isLeapYear(year))) {
+			goto fallback
+		} else if day == 31 {
+			switch month {
+			case 4, 6, 9, 11:
+				goto fallback
```

Comment:
I noticed `time.Parse` generates error messages for the input, e.g. `parsing time "2000-12-31T23:59:60Z": second out of range`. If we didn't care about strict compatibility of error messages, we could create global errors for `{unit} out of range` and return them here to avoid allocations on the error paths.

Reply:
That seems fine to me, we are discarding the errors in the json parser anyways 👍